### PR TITLE
DAML-LF: update spec with scenarios

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
In this PR, we add scenarios to the DAML-LF spec.
This include syntax and type inference rule, but let scenarios interpretation unspecified.

In particular, we make clear that a compliant engine must load any archive that contains scenarios if and only if those are well-typed, even if this engine does not implement interpretation of scenarios. 

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/spec/daml-lf-1.rst [hunk 1]
```diff
@@ -2372,6 +2391,24 @@ cases where sub-expressions fail. Those case can be inferred in a
 straightforward way by following the left-to-right evaluation order.
 
 
+About scenario interpretation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The interpretation of scenarios is a feature an engine can provide to
+test business logic within a DAML-LF archive. Nevertheless, the
+present specification does not define how scenarios should be actually
+interpreted. A engine compliant with this specification does not have
```
Comment (cocreature) [inline daml-lf/spec/daml-lf-1.rst:2400]:
```suggestion
interpreted. An engine compliant with this specification does not have
```

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 2]
```diff
@@ -765,18 +765,10 @@ private[validation] object Typing {
...
 
-    private def typeOfCommit(typ: Type, party: Expr, update: Expr): Type = {
+    private def checkCommit(typ: Type, party: Expr, update: Expr): Type = {
       checkType(typ, KStar)
       checkExpr(party, TParty)
       checkExpr(update, TUpdate(typ))
-      TScenario(typ)
-    }
-
-    private def typeOfMustFailAt(typ: Type, party: Expr, update: Expr): Type = {
```
Comment (cocreature) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:775]: Was this just unused?

File: daml-lf/spec/daml-lf-1.rst [hunk 3]
```diff
@@ -883,90 +894,93 @@ We now formally defined *well-formed types*. ::
...
 
                        ┌───────────────┐
-  Well-formed types    │ Γ  ⊢  τ  :  k │
+ Well-formed types    │ Γ  ⊢  τ  :  k │
                        └───────────────┘
 
-      α : k ∈ Γ
-    ————————————————————————————————————————————— TyVar
-      Γ  ⊢  α  :  k
+     α : k ∈ Γ
```
Comment (cocreature) [inline daml-lf/spec/daml-lf-1.rst:900]: So many unnecessary changes :slightly_frowning_face:

Comment (remyhaemmerle-da) [reply daml-lf/spec/daml-lf-1.rst:900]: github has a nice feature  "Hide whitespace changes"

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:775]: I factorize the comment part of `typeOfCommit` and `typeOfMustFailAt`

Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: I revert this change as it is not related the the concern of the PR.

File: daml-lf/spec/daml-lf-1.rst [hunk 4]
```diff
@@ -3592,6 +3632,10 @@ validation phases can be distinguished.
   abstract syntax tree of the package. It is concerned with the
   `well-formedness <Well-formed packages_>`_ of the package.
 
+An engine compliant with the present specification must accept loading a
+package if and only if the latter these two validation passes.
```
Comment (gerolf-da) [inline daml-lf/spec/daml-lf-1.rst:3636]: Is this missing an `of`?
```suggestion
package if and only if the latter of these two validation passes.
```

File: daml-lf/spec/daml-lf-1.rst [hunk 5]
```diff
@@ -695,13 +697,22 @@ Then we can define our kinds, types, and expressions::
...
+       |  'sbind' x₁ : τ₁ ← e₁ 'in' e₂              -- ScenarioBlock
+       |  'commit' @τ e u                           -- ScenarioCommit
+       |  'must_fail_at' @τ e u                     -- ScenarioMustFailAt
+       |  'pass' e                                  -- ScenarioPass
+       |  'sget_time'                               -- ScenarioGetTime
+       |  'sget_party' e                            -- ScenarioGetParty
+       |  'sembed_expr' @τ e                        -- ScenarioEmbedExpr
 
 .. note:: The explicit syntax for maps (cases ``ExpTextMap`` and
   ``ExpGenMap``) is forbidden in serialized programs. It is specifies
```
Comment (associahedron) [inline daml-lf/spec/daml-lf-1.rst:711]:
```suggestion
  ``ExpGenMap``) is forbidden in serialized programs. It is specified
```
